### DISCUSSION TITLE
Can I use OhmJS to parse binary values?

### DISCUSSION THREAD

**Discussion by nileshtrivedi** [ohmjs/ohm#491] [General]
It seems like OhmJS seems to be designed to parse _textual_ (ascii or UTF-8) languages. But surely, PEG should be able to handle raw binary values too? So is this just a limitation of OhmJS then?

For example, if I want to parse 32-bit IPv4 addresses:

```
IPAddress {
  ipv4address = byte byte byte byte
  byte = bit bit bit bit bit bit bit bit
  bit = ("0" | "1")
}
```

The string `10111001101110011011100110111001` would be parsed successfully but the actual 4-byte value would not.
How do I go about using OhmJS for parsing binary protocols?

**Comment by pdubroy** [comment]
Yes, Ohm is designed for textual languages.

You're right that in general, PEGs could be applied to binary data. But to do something like you're describing here, the parser would need to track the position in the input stream at the level of bits, which Ohm doesn't do.

**Comment by joshmarinacci** [comment]
I remember the precursor to Ohm could do binary data, though it was harder to use and didn't have all of Ohm's niceties. I'd love to see another tool like that one. OMeta, I think it was called.

Josh Marinacci
Recovering Eng

Sent via Superhuman ( ***@***.*** )
